includes/MediaWiki/ReconciliationService.php

```diff
@@ -0,0 +1,144 @@
+	private $entityRevisionLookup;
+
+	/** @var IdGenerator */
+	private $idGenerator;
+
+	/** @var ExternalLinks */
+	private $externalLinks;
+
+	/** @var TitleFactory */
+	private $titleFactory;
+
+	public function __construct(
+		EntityIdLookup $entityIdLookup,
+		EntityLookup $entityLookup,
+		EntityRevisionLookup $entityRevisionLookup,
+		IdGenerator $idGenerator,
+		ExternalLinks $externalLinks,
+		TitleFactory $titleFactory
+	) {
+		$this->entityIdLookup = $entityIdLookup;
+		$this->entityLookup = $entityLookup;
+		$this->entityRevisionLookup = $entityRevisionLookup;
+		$this->idGenerator = $idGenerator;
+		$this->externalLinks = $externalLinks;
+		$this->titleFactory = $titleFactory;
+	}
+
+	/**
+	 * @param int[] $pageIds
+	 * @return ItemId[]
+	 */
+	private function getItemIdsFromPageIds( array $pageIds ) : array {
+		$titles = $this->titleFactory->newFromIDs( $pageIds );
+
+		if ( empty( $titles ) ) {
+			return [];
+		}
+
+		$entityIds = $this->entityIdLookup->getEntityIds( $titles );
+		$itemIds = [];
+		foreach ( $entityIds as $entityId ) {
+			if ( $entityId instanceof ItemId ) {
+				$itemIds[] = $entityId;
+			}
+		}
+		return $itemIds;
+	}
+
+	/**
+	 * Finds existing item or returns a new one if no matching item is found.
+	 * Lookup is matching on items with statement ($reconcileUrlProperty) set to the $reconciliationUrl
+	 *
+	 * @param PropertyId $reconcileUrlProperty
+	 * @param string $reconciliationUrl
+	 * @return ReconciliationItem
+	 * @throws \Exception
+	 */
+	public function getOrCreateItemByStatementUrl(
+		PropertyID $reconcileUrlProperty,
+		string $reconciliationUrl
+	) : ReconciliationItem {
+		// Find Items that use the URL
+		$itemIdsThatReferenceTheUrl = $this->getItemIdsFromPageIds(
+			$this->externalLinks->pageIdsContainingUrl( $reconciliationUrl )
+		);
+
+		// Find Items that match the URL and Property ID
+		$itemsThatReferenceTheUrlInCorrectStatement = [];
+		foreach ( $itemIdsThatReferenceTheUrl as $itemId ) {
+			/** @var Item $item */
+			$item = $this->entityLookup->getEntity( $itemId );
+
+			foreach ( $item->getStatements()->getByPropertyId( $reconcileUrlProperty )->toArray() as $statement ) {
+				if ( !$statement->getMainSnak() instanceof PropertyValueSnak ) {
+					continue;
+				}
+				/** @var PropertyValueSnak $mainSnak */
+				$mainSnak = $statement->getMainSnak();
+
+				$urlOfStatement = $mainSnak->getDataValue()->getValue();
+				if ( $urlOfStatement === $reconciliationUrl ) {
+					$itemsThatReferenceTheUrlInCorrectStatement[] = $item;
+				}
+			}
+		}
+
+		// If we have more than one item matches, something is wrong and we can't edit
+		if ( count( $itemsThatReferenceTheUrlInCorrectStatement ) > 1 ) {
+			die( 'Matched multiple Items during reconciliation :(' );
+		}
+
+		// Get our base
+		if ( count( $itemsThatReferenceTheUrlInCorrectStatement ) === 1 ) {
+			$base = $itemsThatReferenceTheUrlInCorrectStatement[0];
+			// XXX: This bit is so annoying...
```

Comment:
We can probably alleviate this by using an `EntityRevisionLookup` rather than an `EntityLookup` in the earlier block where we load the item, though that should be done in a separate commit.

Reply:
put this in it's own task https://phabricator.wikimedia.org/T282553